### ISSUE TITLE
Suggestion: sum types / tagged union

### ISSUE THREAD

**Issue by alainfrisch** [Microsoft/TypeScript#186]
A very nice addition to TypeScript's type system would be sum types in the spirit of ML-like languages.  This is one of basic and simple programming constructs from functional programming which you really miss once you get used to it, but which seem to have a hard time being included in new modern languages (contrary to other features from functional programming such as first-class functions, structural types, generics).

I guess the most natural way to integrate sum types in the current language syntax would be to extend enum variants with extra parameters (similarly to what Rust does:  http://doc.rust-lang.org/tutorial.html#enums ) and upgrade the switch statement to a more powerful structural pattern matching (although in a first step, simply discriminating on the toplevel variant and capturing its parameters would be already quite good).

This is quite different from other the proposal about "union types"  (#14), which would mostly be useful to capture types in existing Javascript APIs.  Sum types are rather used to describe algebraic data structures. They would be particularly useful for any kind of symbolic processing (including for implementing the TypeScript compiler).

**Comment by danquirk** [comment]
@aleksey-bykov as much as I love the syntax used for this concept in functional languages I think our current type guard concept is a very elegant solution for the problem in TypeScript. It allows us to infer and narrow union cases using existing JavaScript patterns which means more people will benefit from this than if we required people to write new TypeScript specific syntax. Hopefully we can extend type guards further (there are a number of suggestions for that already).

**Comment by aleksey-bykov** [comment]
The bigest benefit of sum types is their propery of exhaustiveness or
totality if you will. Type guards are not going to give it, are they?

**Comment by danquirk** [comment]
That's definitely a big part of their value. It's something I would like to look at but haven't filed a formal issue for yet. Certainly you could imagine an error for something like this:

```ts
function foo(x: string|number) {
    if(typeof x === "string") { ... }
    // error, unhandled case: typeof x === "number"
}
```

the question is whether it could be robust enough to handle more complicated cases.

**Comment by aleksey-bykov** [comment]
Another thing I truly dont understand about type guards is when I do a
check against an interface rather than class (function constructor), say:

```
interface A {}
type T = string | A;
var x : T = undefined;
if (x instanceof A) then alert('Hey!');
```
How is this supposed to work?

**Comment by danquirk** [comment]
Well that wouldn't do what you want in JavaScript either if x was some object literal. Your type guard would either be a set of property checks

```ts
if(x.length) { ... } 
if(x.aPropertyInA) { ... }
```

or we could do something like #1007

**Comment by aleksey-bykov** [comment]
Is checking for property considered a type guard? I mean I know I can sniff
properties just like I would do in plain JavaScript. What my question is if
a check for a property like you put it would give me any tyoe safety in
then block after such check.

**Comment by danquirk** [comment]
Not at the moment, but I implemented it over here to test it out https://github.com/Microsoft/TypeScript/issues/1260. So you can see why I might think there's enough room to extend this functionality far enough to get the full breadth of checking you're imagining without needing new syntax (which means more people benefit and we make life easier on ourselves if that syntax is needed by ES later).

**Comment by aleksey-bykov** [comment]
I am all up for keeping as less new syntax as possible. If you can fit a
new feature in old syntax it's good for everyone. Then I will withdraw my
proposal as needless.

**Comment by alainfrisch** [comment]
> The bigest benefit of sum types is their propery of exhaustiveness or totality if you will. Type guards are not going to give it, are they?

Nevor's proposal for implementing sum types through literal singleton types (https://github.com/Microsoft/TypeScript/issues/1003) supports checking exhaustiveness.  After exhausting all cases through type guards, the type is reduced to the empty type (void), and you can ask the compiler to check for exhaustiveness by adding a final call to the "impossible" function defined as `function impossible(v : void) { throw "impossible"; }`

**Comment by aleksey-bykov** [comment]
I cannot see how a exhaustiveness can be enforced this way. What if I simply missed a case in a switch? Will compiler give me an error? For example:
```
type Ready = { value : "ready" };
type NotReady = { value : "notready" };
type Value = Ready | NotReady;
var v : Value;
switch (v.value) {
   case "ready":
     break;
   // TODO: Don't forget to uncomment the following!
   // case "notready":
   //  break;
   default:
      impossible(v); // okay as long as the type is not extended
}
```
An error from a compiler about missing case of "notready" is required in order to call it "exhaustiveness".

But even in case there was an error it would be a breaking change, because the switch operator has never been exhaustive.

With all this said it doesn't seem to be that exhaustiveness can be enforced without additional syntax.

**Comment by alainfrisch** [comment]
Yes, you get exhaustiveness check.  If you forget the "notready" case, then the default branch is type-checked asssuming that v has type NotReady (which is what remains from Ready|NotReady after you matched on "ready"), and type-checking of that default branch fails, because impossible expects that v has type void (and NotReady is not a subtype of void).

And it's not a breaking change, because the error is only triggerred by the call to the impossible function.  If you don't call it, you don't get the error (and you don't get the exhaustiveness check).  This is the only weakness here: you need to think about adding the default branch and call impossible in it or you don't get the exhaustiveness check.

**Comment by aleksey-bykov** [comment]
Oh, now I see what you are saying. The part I didn't realize was *"impossible expects that v has type void"*

**Comment by fdecampredon** [comment]
Honestly even if we have property type-guard the syntax will be somewhat very unnatural and verbose
```typescript
enum ColorEnum {
  Red,
  Green,
  Blue,
  Rgb,
  Rgba,
}
type Color = 
  { value: ColorEnum.RED } | 
  { value: ColorEnum.GREEN } | 
  { value: ColorEnum.Blue} | 
  { value: ColorEnum.Rgb , r: number; g : number, b: number } |
  { value:  ColorEnum.Rgba, r: number; g : number, b: number, a: number }

function logColor(color: Color) {
  switch(color.value) {
    case ColorEnum.Rgb : 
      var {r, g, b} = color;
      console.log(....);
      break;
    case ColorEnum.Rgba:
      var {r, g, b, a} = color;
      console.log(....);
      break;
    default: 
      ....
}
```

Also in this kind of construct, we can seriously greatly feel that the language could provide us shorhand .. 

Something like: 

```typescript
type Color = ?keyword? {
  Red,
  Green,
  Blue,
  Rgb(r: number, g:  number, b:  number ),
  Rgba(r: number, g:  number, b:  number, a: number)
}

function logColor(color) {
  match(color) {
    ColorEnum.Rgb(r, g, b):
      console.log(....);
      break;
    ColorEnum.Rgba(r, g, b, a):
      console.log(....);
      break;
    default: 
      ....
}
```

would make more sense fore me, but sure it creates a lot of new syntax

**Comment by jbondc** [comment]
Didn't think was possible unless you'd add metadata. But after playing with type guards and #1003 would be pretty cool. 

e.g.
```javascript
enum ColorEnum {
  Red,
  Green,
  Blue
}

// This will have to emit 'Rgb' and 'Rgba' as JS classes, better naming?
type class ColorApi {
  Rgb(r: number, g:  number, b:  number )
  Rgba(r: number, g:  number, b:  number )
}

type red = { red : number };
type blue = {  blue : number };
type yellow = { yellow: number };
type ColorStruct = red | blue | yellow
type ColorNames = "red" | "blue" | "yellow"

// sum type, forces use of type guards to exhaust all possible cases
type Color = ColorEnum + ColorApi + ColorStruct + ColorNames + string

function logColor(color: Color) {
  if(color == null) {
    // null or undefined, narrows to void type
  } elseif(color instanceof ColorEnum) {
  } elseif(color instanceof Rgb) {
  } elseif(color instanceof Rgba) {
  } elseif(typeof color.red === "number" || typeof color.blue === "number" || typeof color.yellow === "number") {
  } elseif(color === "red" || color === "blue" || color === 'yellow') {
  } elseif(typeof color === 'string') {
      console.log("convert this from hex?")
  } else {
     throw ?
  }
}
```
Makes good candidate for a test case :)

**Comment by jbondc** [comment]
Actually the 'ColorApi' part probably may not work well , probably the hardest :/

**Comment by donaldpipowitch** [comment]
Sum types/tagged union like Rusts would be a really great addition. I hope this is something which can land in the foreseeable future, so it maybe can influence other discussions about types in JavaScript (like the recent SoundScript).

**Comment by roganov** [comment]
I would really love this feature.
For as long as it's not implemented, what is the recommended way to emulate the following Elm snippet?
```elm
type Action
    = NoOp
    | UpdateField String
    | EditingTask Int Bool
    | UpdateTask Int String
```
@DanielRosenwasser has kindly shown me a pretty cool way (comment [#4190](https://github.com/Microsoft/TypeScript/issues/4180#issuecomment-128265521)) way to achieve what I want, but unfortunately it requires quite a lot of boilerplate code. In a mid-sized flux application there might be hundreds of actions so it isn't very convenient.

**Comment by acjay** [comment]
I'm really hoping we get this. Javascript is already a very functional language, and some sort of support for ADTs would really give it much of the additional power that all the typed-functional programming languages possess with their pattern matching. I noticed that some of the proposals involved storing tags on objects. This could be made less brittle with ES6 symbols.

**Comment by Zorgatone** [comment]
You could set the prototype's `valueOf` method, and then use Math operators on it in JavaScript.

The question is if TypeScript will allow that too, or you get an error (with/without classes)

**Comment by andy-hanson** [comment]
It would be nice to have a `sealed` keyword on an abstract class to indicate that the only subclasses intended to exist are those defined in the same project.

That way you could do this:

	sealed abstract class Super {}
	class Sub1 extends Super {}
	class Sub2 extends Super {}

	function f(s: Super) {
		if (s instanceof Sub1) {

		} else {
			// s is a Sub2
		}
	}

**Comment by ivogabe** [comment]
@andy-hanson You can do that with a union type:
```typescript
type Super = Sub1 | Sub2;
class Sub1 { a: string; }
class Sub2 { b: string; }
function (s: Super) {
  if (s instanceof Sub1) {
    // s: Sub1
  } else {
    // s: Sub2
  }
}
```

**Comment by andy-hanson** [comment]
@ivogabe The idea is that one should be able to define the class (with methods) and the type as the same thing, rather than having separate `class Super` and `type SuperT = Sub1 | Sub2`.

**Comment by roganov** [comment]
@andy-hanson Can't you just override required methods on `Sub1` and `Sub2` (in other words, dynamic dispatch)? Why would you want to check types?

**Comment by opensrcken** [comment]
If I'm not mistaken, ScalaJs has to handle a similar concept.

```scala
abstract class Exp
 
case class Fun(e: Exp) extends Exp
case class Number(n: Int) extends Exp
case class Sum(exp1: Exp, exp2: Exp) extends Exp
case class Product(exp1: Exp, exp2: Exp) extends Exp

def print(e: Exp): String = e match {
    case Number(1) => "1"
    case Number(x) => x.toString
    case Sum(Number(1), Number(2)) => "(1 + 2)"
    case Sum(e1, e2) => "(+ " + print(e1) + " " + print(e2) + ")"
    case Product(e1, e2) => "(* " + print(e1) + " " + print(e2) + ")"
    case Fun(e) => "(fn [] " + print(e) + ")"
}
```

compiles to

```javascript
$c_Ltutorial_webapp_TutorialApp$.prototype.print__Ltutorial_webapp_TutorialApp$Exp__T = (function(e) {
  var rc19 = false;
  var x2 = null;
  var rc20 = false;
  var x5 = null;
  if ($is_Ltutorial_webapp_TutorialApp$Number(e)) {
    rc19 = true;
    x2 = $as_Ltutorial_webapp_TutorialApp$Number(e);
    var p3 = x2.n$2;
    if ((p3 === 1)) {
      return "1"
    }
  };
  if (rc19) {
    var x = x2.n$2;
    return ("" + x)
  };
  if ($is_Ltutorial_webapp_TutorialApp$Sum(e)) {
    rc20 = true;
    x5 = $as_Ltutorial_webapp_TutorialApp$Sum(e);
    var p6 = x5.exp1$2;
    var p7 = x5.exp2$2;
    if ($is_Ltutorial_webapp_TutorialApp$Number(p6)) {
      var x8 = $as_Ltutorial_webapp_TutorialApp$Number(p6);
      var p9 = x8.n$2;
      if ((p9 === 1)) {
        if ($is_Ltutorial_webapp_TutorialApp$Number(p7)) {
          var x10 = $as_Ltutorial_webapp_TutorialApp$Number(p7);
          var p11 = x10.n$2;
          if ((p11 === 2)) {
            return "(1 + 2)"
          }
        }
      }
    }
  };
  if (rc20) {
    var e1 = x5.exp1$2;
    var e2 = x5.exp2$2;
    return (((("(+ " + this.print__Ltutorial_webapp_TutorialApp$Exp__T(e1)) + " ") + this.print__Ltutorial_webapp_TutorialApp$Exp__T(e2)) + ")")
  };
  if ($is_Ltutorial_webapp_TutorialApp$Product(e)) {
    var x13 = $as_Ltutorial_webapp_TutorialApp$Product(e);
    var e1$2 = x13.exp1$2;
    var e2$2 = x13.exp2$2;
    return (((("(* " + this.print__Ltutorial_webapp_TutorialApp$Exp__T(e1$2)) + " ") + this.print__Ltutorial_webapp_TutorialApp$Exp__T(e2$2)) + ")")
  };
  if ($is_Ltutorial_webapp_TutorialApp$Fun(e)) {
    var x14 = $as_Ltutorial_webapp_TutorialApp$Fun(e);
    var e$2 = x14.e$2;
    return (("(fn [] " + this.print__Ltutorial_webapp_TutorialApp$Exp__T(e$2)) + ")")
  };
  throw new $c_s_MatchError().init___O(e)
});

// type check for Number class
function $is_Ltutorial_webapp_TutorialApp$Number(obj) {
  return (!(!((obj && obj.$classData) && obj.$classData.ancestors.Ltutorial_webapp_TutorialApp$Number)))
}
```

**Comment by andy-hanson** [comment]
@roganov: Many people prefer a style of programming where functions are defined only once, as opposed to once per subclass. That's why this issue here exists.

**Comment by Artazor** [comment]
Since we have string literal types, I would propose to interpret the following declaration
```TypeScript
interface Action[type] {   // the name "type" will be used as discriminator tag
     "REQUEST": {}
     "SUCCESS": { data: string }
     "FAILURE": { error: any } 
}
```
as **fully equivalent** shorthand to
```TypeScript
type Action = { type: "REQUEST" } 
            | { type: "SUCCESS", data: string } 
            | { type: "FAILURE", error: any }
```

and **detect** such situation in `if` and `switch` in the following manner

**if all** conditions are met:

* expression of form `expr.prop` is compared against expression `S` for equality or non-equuality; 
* syntactically `expr` is simple identifier (no compound expressions); 
* the type of the `S` expression is singleton string literal;
* expression `expr` itself has type that in each union branch has the same property named `prop` with singleton string literal type (not necessarily disjoint across the union branches)

**then** in the proper following context the type of `expr` is narrowed to the union of the branches where type of the `prop` field exactly matches the type of `S` (or doesn't match if the check was for non-equality), or type: `{ prop: S }` if no alternatives found. 

Thus we will be able to write

```TypeScript
var a: Action;
// ...
if (a.type === "SUCCESS") {
     console.log(a.data) // type of a is narrowed to {type: "SUCCESS", data: string}
}
```
This proposal is less powerfull than full algebraic types (it lacks recursion), nevertheless it's rather pragmatic as it helps to assign the correct types to commonly used patterns in JavaScript especially for React + (Flux/Redux).

I know that narrowing is working only on simple variables, but I think this case is somewhat ideologically equivalent to the type-checking for expressions assigned to the simple variable.

@danquirk, what is your opinion?

**Comment by weswigham** [comment]
@Artazor I've already done some [investigation](https://github.com/Microsoft/TypeScript/pull/6062) into equality -based type narrowing - It is pretty cool to use.

**Comment by DanielRosenwasser** [comment]
@Artazor with @weswigham's type narrowing branch, and some recent changes I've made, that works. It needs to happen one step at a time though.

**Comment by simonbuchan** [comment]
If you're looking for matching sugar for using existing types, maybe allow destructuring in cases?:

```
function reducer(s: MyImmutableState, a: Action): MyImmutableState {
    switch (a) {
        case {type: "REQUEST"}:
            return s.merge({spinner: true});
        case {type: "SUCCESS", data}:
            return s.merge({spinner: false, data, error: null});  
        case {type: "FAILURE", error}:
            return s.merge({spinner: false, data: null, error});
        default:
            return s;
    }
}
```

This may already have a (nonsense) meaning of check if `a` is the same instance of this new literal. Other options:
- `case {type is "FAILURE", error}:', suggesting type guarding is happening, but they are not emitted.
- `case let {type: "FAILURE", error}:`, though I'm concerned about it being too easy to miss the `let`.
- `case {type == "FAILURE", error}:`, which would allow other destructurings like `{opCount == 2, op1, op2}`, `{opCount > 3, opList}`, but `type = "FAILURE"` (meaning default missing `type` in existing ES6) would, again be too close.

Alternatively: Perhaps depend on adding C++-style condition declarations (`if (let foo = getFooOrNull()) foo.bar();`, and allow:

```
if (let {type == "FAILURE", error} = a) ...
```

I think this is closer to how languages with pattern matching work in a way that feels like where ES is going, but it seems like (something like) it should be proposed as an ES feature first.

**Comment by basarat** [comment]
FWIW @Artazor's proposal https://github.com/Microsoft/TypeScript/issues/186#issuecomment-173719173 is what flow does (disjoint unions) : http://flowtype.org/docs/disjoint-unions.html#_

```ts
type Action = { type: "REQUEST" } 
            | { type: "SUCCESS", data: string } 
            | { type: "FAILURE", error: any }
```
Conditional checks on `type` allow you to narrow down the other members of `Action` :rose:

**Comment by weswigham** [comment]
@Basarat I have a version of our narrowing code which allows this behavior
with existing types (#6062) - but it still has some open questions.

**Comment by ahejlsberg** [comment]
Implementation of discriminated union types using string literal type tags is now available in #9163.

**Comment by shelby3** [comment]
@aleksey-bykov [wrote](https://github.com/Microsoft/TypeScript/issues/186#issuecomment-68818764):
> Another thing I truly dont understand about type guards is when I do a
check against an `interface` ***rather than*** `class` (function constructor), say:

Use an [`abstract class` instead of `interface`](https://github.com/Microsoft/TypeScript/issues/1007#issuecomment-246085962).

@danquirk [wrote](https://github.com/Microsoft/TypeScript/issues/186#issuecomment-68818961):
> Well that wouldn't do what you want in JavaScript either if x was some object literal. Your type guard would either be a set of property checks
>
> ```
> if(x.length) { ... } 
> if(x.aPropertyInA) { ... }
> ```

That is a structural guard which may not always be applicable, thus my nominal suggestion above to use `abstract class`.

> or we could do something like #1007 

Which [appears to me to be undesirable](https://github.com/Microsoft/TypeScript/issues/1007#issuecomment-246085962).
